packages/core-data/src/locks/test/engine.js

```diff
@@ -0,0 +1,135 @@
+
+		// on each grant, verify that the other lock is not granted at the same time
+		const check1 = l1.then( () => {
+			l1Granted = true;
+			expect( l2Granted ).toBe( false );
+		} );
+
+		const check2 = l2.then( () => {
+			l2Granted = true;
+			expect( l1Granted ).toBe( false );
+		} );
+
+		// unlock both
+		const lock1 = await l1;
+		locks.release( lock1 );
+		l1Granted = false;
+
+		const lock2 = await l2;
+		locks.release( lock2 );
+		l2Granted = false;
+
+		// ensure that both locks were granted and checked
+		return await Promise.all( [ check1, check2 ] );
+	} );
+
+	it( 'does not grant an exclusive lock if a non-exclusive one already exists', async () => {
+		const locks = createLocks();
+
+		let l1Granted = false;
+		let l2Granted = false;
+
+		// request two locks
+		const l1 = locks.acquire( 'store', [ 'root' ], false );
+		const l2 = locks.acquire( 'store', [ 'root' ], true );
+
+		// on each grant, verify that the other lock is not granted at the same time
+		const check1 = l1.then( () => {
+			l1Granted = true;
+			expect( l2Granted ).toBe( false );
+		} );
+
+		const check2 = l2.then( () => {
+			l2Granted = true;
+			expect( l1Granted ).toBe( false );
+		} );
+
+		// unlock both
+		const lock1 = await l1;
+		locks.release( lock1 );
+		l1Granted = false;
+
+		const lock2 = await l2;
+		locks.release( lock2 );
+		l2Granted = false;
+
+		// ensure that both locks were granted and checked
+		return await Promise.all( [ check1, check2 ] );
+	} );
+
+	it( 'does not grant two exclusive locks to parent and child', async () => {
+		const locks = createLocks();
+
+		let l1Granted = false;
+		let l2Granted = false;
+
+		// request two locks
+		const l1 = locks.acquire( 'store', [ 'root' ], true );
+		const l2 = locks.acquire( 'store', [ 'root', 'child' ], true );
+
+		// on each grant, verify that the other lock is not granted at the same time
+		const check1 = l1.then( () => {
+			l1Granted = true;
+			expect( l2Granted ).toBe( false );
+		} );
+
+		const check2 = l2.then( () => {
+			l2Granted = true;
+			expect( l1Granted ).toBe( false );
+		} );
+
+		// unlock both
+		const lock1 = await l1;
+		locks.release( lock1 );
+		l1Granted = false;
+
+		const lock2 = await l2;
+		locks.release( lock2 );
+		l2Granted = false;
+
+		// ensure that both locks were granted and checked
+		return await Promise.all( [ check1, check2 ] );
+	} );
+
+	it( 'grants two non-exclusive locks at once', async () => {
+		const locks = createLocks();
+
+		const l1 = await locks.acquire( 'store', [ 'root' ], false );
+		const l2 = await locks.acquire( 'store', [ 'root' ], false );
+
+		expect( l1 ).not.toBeUndefined();
+		expect( l2 ).not.toBeUndefined();
+	} );
+
+	it( 'grants two exclusive locks to different branches', async () => {
+		const locks = createLocks();
+
+		const l1 = await locks.acquire( 'store', [ 'a' ], true );
+		const l2 = await locks.acquire( 'store', [ 'b' ], true );
+
+		expect( l1 ).not.toBeUndefined();
+		expect( l2 ).not.toBeUndefined();
```

Comment:
```suggestion

		await expect(
			locks.acquire( 'store', [ 'a' ], true )
		).resolves.not.toBeUndefined();
		await expect(
			locks.acquire( 'store', [ 'b' ], true )
		).resolves.not.toBeUndefined();
```

Reply:
Similarly, other tests could also use `.resolves` / `.rejects`

Reply:
The only reason these `expect` calls are there is because it's required that a unit test has at least one assertion. For this unit test, success means that it finishes without throwing an exception or getting stuck in a deadlock.

For example, when trying to acquire the same exclusive lock twice:
```js
await locks.acquire( 'store', [ 'a' ], true );
await locks.acquire( 'store', [ 'a' ], true );
```
the second `await` should never return.

Reply:
I'm not sure I follow – the red part and the green part of the suggestion are equivalent and they're both assertions, just formulated a bit differently.

The upside of separate `expect()` calls though is that you may use a very similar structure when testing the opposite behavior – I don't think jest has `.not.resolves`.

Reply:
I just prefer the original version, with the async flow (`await locks.acquire()`) separate from the asserts. The async flow is quite important here, so it shouldn't be hidden inside `expect()` calls.

Reply:
Sure! It's certainly a matter of taste.